### DISCUSSION TITLE
Matching an enum that contains a String

### DISCUSSION THREAD

**Discussion by bigyihsuan** [kevinmehall/rust-peg#335] [Q&A] [answered]
I want a rule that should match specifically the value `Token::Element("H".to_string())`, and a rule that matches every other string. The `Token` enum has a variant `Token::Element(String)`. What I have currently is the following:

```rs
pub rule element<S: Symbol, B: Backend<S>>(interner: &mut StringInterner<B>) -> Element<S>
    = [Token::Element(e)]
    { let e = interner.get_or_intern(e); Element(e) } // ok

pub rule hydrogen<S: Symbol, B: Backend<S>>(interner: &mut StringInterner<B>) -> Element<S>
    = [Token::Element("H".to_string())] // error here
    { let h = interner.get_or_intern("H"); Element(h) }
```

However, I get the following errors from the compiler:

```rs
cannot find tuple struct or tuple variant `to_string` in this scope
not found in this scope rustc[E0531]

this pattern has 2 fields, but the corresponding tuple variant has 1 field
expected 1 field, found 2 rustc[E0023]

tok.rs(52, 13): tuple variant has 1 field // this is correct, Token::Element only has 1 field
```

How can I write a rule that specifically matches `Token::Element("H".to_string())` and nothing else?

**Comment by kevinmehall** [comment, marked as answer]
The `[]` syntax expands into a the pattern of a `match` statement like

```rust
match token {
    Token::Element("H".to_string()) => { let h = interner.get_or_intern("H"); Element(h) }
     ...
}
```
which produces the same error because the `.to_string()` is an expression, but not valid as a pattern on the left hand side of a match.

There's a [Rust proposal to make `match` patterns dereference](https://github.com/rust-lang/rust/issues/87121) a String to into a `str` that you can match with a string literal pattern like `"H"`. Looks like it may even be [implemented](https://github.com/rust-lang/rust/pull/98914) for String already if you turn on `#![feature(string_deref_patterns)]` on Nightly.

What you can do on stable Rust right now though is use an [`if` guard](https://doc.rust-lang.org/reference/expressions/match-expr.html#match-guards), which rust-peg passes through onto the match arm. The part after the `if` is an expression returning a boolean that can access variables bound in the pattern.

```rust
[Token::Element(e) if e == "H"]
```

**Reply by bigyihsuan** [reply]
It didn't occur to me that `[]` is just a match pattern, this is extremely helpful info! Thanks!
